solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/BeforeReconnect.java

```diff
@@ -16,6 +16,17 @@
  */
 package org.apache.solr.common.cloud;
 
-public interface BeforeReconnect {
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.curator.framework.state.ConnectionStateListener;
+
+public interface BeforeReconnect extends ConnectionStateListener {
   public void command();
+
+  @Override
+  default void stateChanged(CuratorFramework client, ConnectionState newState) {
```

Comment:
I really don't know if this is right. Its just like OnDisconnect which seems wrong but tests pass :/

Reply:
I'm not really sure about any of these classes. It seems like `OnDisconnect` should do `!newState.isConnected()` instead of `newState == ConnectionState.LOST`... I have a slight memory about doing this stuff and there being reasons, but I can't remember why I did this... It definitely warrants a much more thorough look. I'm not even sure having a `BeforeReconnect` is a good idea. The curator APIs don't really give us a good way of implementing it.